File: sdk/python/tests/unit/test_sensitive.py
import json

import pytest

import anyvali as v


def test_sensitive_data_round_trip_and_validation():
    schema = v.object_({
        "name": v.string(),
        "secret": v.string().min_length(3).describe("secret", sensitive=True),
        "profile": v.object_({"token": v.string()}).describe("profile", sensitive=True),
        "aliases": v.array(v.string().describe("alias", sensitive=True)),
        "note": v.nullable(v.string().describe("note", sensitive=True)),
    })
    plain = {
        "name": "Ada",
        "secret": "abc",
        "profile": {"token": "xyz"},
        "aliases": ["one", "two"],
        "note": None,
    }

    seen = []

    def enc(path, value):
        seen.append(path)
        return "encrypted:" + json.dumps(value, separators=(",", ":"))

    encrypted = v.encrypt(schema, plain, enc)
    assert encrypted == {
        "name": "Ada",
        "secret": 'encrypted:"abc"',
        "profile": 'encrypted:{"token":"xyz"}',
        "aliases": ['encrypted:"one"', 'encrypted:"two"'],
        "note": None,
    }
    assert seen == [["secret"], ["profile"], ["aliases", 0], ["aliases", 1]]
    assert v.safe_parse_encrypted(schema, encrypted).success
    assert v.decrypt(schema, encrypted, lambda _path, value: json.loads(value[10:])) == plain

    assert not v.safe_parse_encrypted(schema, {**plain, "secret": "abc"}).success
    with pytest.raises(v.ValidationError):
        v.encrypt(schema, plain, lambda _path, _value: "broken")

File: sdk/ruby/test/test_sensitive.rb
# frozen_string_literal: true

require_relative "test_helper"
require "json"

class SensitiveTest < Minitest::Test
  def test_encrypted_round_trip_and_validation
    schema = AnyVali.object(
      properties: {
        "name" => AnyVali.string,
        "secret" => AnyVali.string.describe("secret", sensitive: true),
        "profile" => AnyVali.object(
          properties: { "token" => AnyVali.string },
          required: ["token"]
        ).describe("profile", sensitive: true),
        "aliases" => AnyVali.array(AnyVali.string.describe("alias", sensitive: true))
      },
      required: %w[name secret profile aliases]
    )
    plain = {
      "name" => "Ada", "secret" => "abc",
      "profile" => { "token" => "xyz" }, "aliases" => %w[one two]
    }

    encrypted = AnyVali.encrypt(schema, plain, ->(_path, value) { "encrypted:#{JSON.generate(value)}" })
    assert AnyVali.safe_parse_encrypted(schema, encrypted).success?
    assert_equal plain, AnyVali.decrypt(
      schema, encrypted, ->(_path, value) { JSON.parse(value.delete_prefix("encrypted:")) }
    )
    assert_raises(AnyVali::ValidationError) do
      AnyVali.encrypt(schema, plain, ->(*) { "broken" })
    end
  end
end
